### PR TITLE
Rework JS manual instrumentation docs

### PR DESCRIPTION
**Pull request by svrnm**
Goal of this PR is to rework the manual instrumentation docs significantly:

* Make use of the "roll the dice" application to have an application to follow along & try things out
* Split out signal-independent setup instructions (dependencies, resource attributes setup)
* Add a lot of text, examples, etc.

I raise this as a *draft* PR first, please provide your feedback if this is going in the right direction, there is some more cleanup & functional testing that needs to be done.

---

**Preview**: https://deploy-preview-2965--opentelemetry.netlify.app/docs/instrumentation/js/manual/

### CODE REVIEW

**Comment by cartermp** [comment]
I like the overall concept a lot.

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:201]:
```suggestion
When using the OpenTelemetry SDK, certain [resource attributes](/docs/instrumentation/js/resources/) are necessary, and beneficial, to provide:
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:205]:
```suggestion
- `service.version` (optional): version of the service API or implementation.
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:210]:
```suggestion
To set these attributes for the example app, create a file named `instrumentation.ts` (or
`instrumentation.js` if not using typescript) and add the following code to it:
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:249]:
```suggestion
Now run the app by requiring the library:
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:266]:
```suggestion
This basic setup has no effect on your app yet. You need to 
add code for [traces](#traces), [metrics](#metrics), and/or [logs](#logs).
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:272]:
```suggestion
{{% alert title="Note" color="info" %}} If you’re instrumenting a library, **skip this step**.
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:275]:
```suggestion
To enable [tracing](/docs/concepts/signals/traces/) in your app, you'll need
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:282]:
```suggestion
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:284]:
```suggestion
As explained next, modify the `instrumentation.*` file to include all the SDK initialization code in Node and the browser.
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:428]:
```suggestion
This will have no effect on your app yet: you need to
[create spans](#create-spans) to have telemetry emitted by your app.
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:357]:
```suggestion
This will have no effect on your app yet: you need to
[create spans](#create-spans) to have telemetry emitted by your app.
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:448]:
```suggestion
#### Configure an exporter
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:454]:
```suggestion
The `ConsoleSpanExporter` will write spans to the console. After you have
finished setting up manual instrumentation, you need to configure
an appropriate exporter to
[export the app's telemetry data](/docs/instrumentation/js/exporters/)
to one or more telemetry backends.
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:492]:
```suggestion
The values of `instrumentation-scope-name` and `instrumentation-scope-version`
should uniquely identify the
[instrumentation scope](/docs/specs/otel/glossary/#instrumentation-scope), such
as the package, module or class name. While the name is required, the version is
optional.
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:499]:
```suggestion
In the case of the [example app](#example-app), there are two
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:626]: This doesn't look good:

> image

How about:

```suggestion
Now that you have [tracers](/docs/concepts/signals/traces/#tracer)
initialized, you can create [spans](/docs/concepts/signals/traces/#spans).
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:629]:
```suggestion
The code below illustrates how to create an active span, which is the most
common kind of span.
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:676]: Is there any reason for readers to not have "followed the instructions ... up to this point"? Isn't that the basic assumption in the page, that by defaul the reader will be using he example app?

```suggestion
If you followed the instructions using the
[example app](#example-app) up to this point, you can copy the
code above in your library file `dice.ts` (or `dice.js`). You should now be
able to see spans emitted from your app.
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:679]:
```suggestion
Start your app as follows, and then send it requests by
visiting <http://localhost:8080/rolldice?rolls=12> with your browser or `curl`.
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:696]:
```suggestion
After a while, you should see the spans printed in the console by the
`ConsoleSpanExporter`, something like this:
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:719]:
```suggestion
nested in nature. For example, the `rollOnce()` function below represents a nested
operation. The following sample creates a nested span that tracks `rollOnce()`:
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:774]:
```suggestion
This code creates a child span for each _roll_ that has `parentSpan`'s
ID as their parent ID:
```

**Review by chalin — COMMENTED** [review]
Looking good. Made a first pass of comments.

File: content/en/docs/instrumentation/js/manual.md
Comment (cartermp) [inline content/en/docs/instrumentation/js/manual.md:455]: Reading this page, I think we could do well to have a code sample.

First, elaborate that the default exporter is `localhost:4317`, which means that if they also configure a collector (link to this...somehow), then it will send to that collector by default.

Then, explicitly say that you're wiring up the console exporter for the purposes of local development/seeing the instrumentation easily. And also include a comment in the source code that says it's for local development and debugging only.

File: content/en/docs/instrumentation/js/manual.md
Comment (cartermp) [inline content/en/docs/instrumentation/js/manual.md:499]:
```suggestion
as the package, module or class name. While the name is required, the version is
still recommended despite being optional.
```

File: content/en/docs/instrumentation/js/manual.md
Comment (cartermp) [inline content/en/docs/instrumentation/js/manual.md:725]: Recommend quickly elaborating that in JS/TS, you need to use `startActiveSpan` here, otherwise you'll have to use the context API to explicitly set spans as active in the current context.

...which is a pain, hence the API! ...but that can be left out :)

File: content/en/docs/instrumentation/js/manual.md
Comment (cartermp) [inline content/en/docs/instrumentation/js/manual.md:903]: Recommend adding a short paragraph at the end stating that the key difference here is `startSpan`, which by default creates an independent span rather than a nested/child span.

**Review by cartermp — COMMENTED** [review]
Overall looking very good!

File: content/en/docs/instrumentation/js/manual.md
Comment (svrnm) [reply content/en/docs/instrumentation/js/manual.md:455]: > First, elaborate that the default exporter is localhost:4317, which means that if they also configure a collector (link to this...somehow), then it will send to that collector by default.

I tried to verify this, but it's actually not the case, here's my instrumentation.ts:

```js
/*instrumentation.ts*/
import { BatchSpanProcessor, ConsoleSpanExporter } from "@opentelemetry/sdk-trace-base";
import { Resource } from "@opentelemetry/resources";
import { SemanticResourceAttributes } from "@opentelemetry/semantic-conventions";
import { NodeTracerProvider } from "@opentelemetry/sdk-trace-node";

const resource =
  Resource.default().merge(
    new Resource({
      [SemanticResourceAttributes.SERVICE_NAME]: "dice-server",
      [SemanticResourceAttributes.SERVICE_VERSION]: "0.1.0",
    })
  );

const provider = new NodeTracerProvider({
    resource: resource,
});
// const exporter = new ConsoleSpanExporter();
// const processor = new BatchSpanProcessor(exporter);
// provider.addSpanProcessor(processor);

provider.register();
```

I also ran it with debugging on, but I don't see any traces being exported. I know that this is true if you use the NodeSDK setup steps (`const sdk = new NodeSDK({ ... }); sdk.start()`).

File: content/en/docs/instrumentation/js/manual.md
Comment (svrnm) [reply content/en/docs/instrumentation/js/manual.md:903]: not sure what your comment is referring to? In the lines above no `startSpan` is used

File: content/en/docs/instrumentation/js/manual.md
Comment (svrnm) [reply content/en/docs/instrumentation/js/manual.md:737]: I added a comment about that a little bit up in the "Create Spans" section:

https://github.com/open-telemetry/opentelemetry.io/pull/2965/files#diff-d5f10bf675fa9b56597823729a28aef6cddf15319e74e6518a546c09297d9021R635

File: content/en/docs/instrumentation/js/manual.md
Comment (svrnm) [reply content/en/docs/instrumentation/js/manual.md:455]: Wondering if the setup section should work with NodeSDK and only comment on the capability to do low-level wiring yourself 🤔 

cc @open-telemetry/javascript-approvers

Comment (cartermp) [reply content/en/docs/instrumentation/js/manual.md:455]: Hrm, I guess I was thinking in terms of the NodeSDK. I think we should finally switch to the Node SDK for this page as well. It was less stable when I initially did this section, but it's been out long enough that we should just do the switch.

**Review by cartermp — COMMENTED** [review]
Okay, re-review -- I think this looks good overall. I think we should finally switch over to the NodeSDK. It aligns with getting-started and it's just so much nicer to use. Plus it makes the exporter bit make sense.

**Comment by svrnm** [comment]
> Okay, re-review -- I think this looks good overall. 

Thanks!

> I think we should finally switch over to the NodeSDK. It aligns with getting-started and it's just so much nicer to use. Plus it makes the exporter bit make sense.

I rewrote a big chunk of that with https://github.com/open-telemetry/opentelemetry.io/pull/2965/commits/04857380b806d89c0614d9542594fdd0b31515db. I kept the other ways of initializing the TracerProvider / MeterProvider in the docs for completeness, not sure if this makes things more complicated or not, wdyt?

Note: I have not tested out the code samples yet in detail, just copied them over, will validate them, right now I am more curious if the structure of the text makes sense so far

**Comment by cartermp** [comment]
> not sure if this makes things more complicated or not, wdyt?

Hmmm, I think it it does. I could see someone getting confused and trying to put this code in their existing Node SDK initialization stuff. I think we should really just do one or the other, with a preference towards the Node SDK. But I guess @open-telemetry/javascript-approvers could have a final say?

File: content/en/docs/instrumentation/js/manual.md
Comment (pichlermarc) [reply content/en/docs/instrumentation/js/manual.md]: Sorry for the late response: As much as it pains me to say, I'm always hesitant to recommend the `NodeSDK` as it has grown over the years and was never fully thought through. It needs a refactor/rewrite, is missing many features one would expect, and should not be considered stable (as indicated by the `0.x` version). It also pulls in a lot of dependencies that cause trouble in certain enviornments (such as the jaeger exporter). 

Therefore, until we address those issues, I think making people depend on its behavior may be doing more harm than good in the long term.

Comment (svrnm) [reply content/en/docs/instrumentation/js/manual.md]: thanks @pichlermarc. I was not aware of this status of `NodeSDK`. We also use it through out the getting started docs for JavaScript and a few other places (can collect them as needed). So this sounds we should keep things as they are with the non-NodeSDK-approach?

Comment (cartermp) [reply content/en/docs/instrumentation/js/manual.md]: How likely are proposed changes to be a big deal? In practice the NodeSDK has been pretty stable, so much so that we (hoenycomb) have been able to rely on it for almost a year in our own distribution. There was a breaking change at one point but it was really simple to update.

Comment (cartermp) [reply content/en/docs/instrumentation/js/manual.md]: I guess this might be worth a SIG convo, but I personally think that's fine enough to document. I would wager the large majority of usage is just in passing in config options and running `start`/`shutdown`.

Comment (pichlermarc) [reply content/en/docs/instrumentation/js/manual.md]: I'll bring it up at this week's JS SIG 🙂 - I think as long as only constructor/start and shutdown are used in the docs this should be fine. 

If the others in the JS SIG meeting on Wednesday agree, I'll set steps to mark these methods as deprecated, then remove them in a later release.

Comment (cartermp) [reply content/en/docs/instrumentation/js/manual.md]: Sounds good! I would also say that if there's some known "this will break people" work that you're planning on, if that work being done helps get us documenting the NodeSDK, then I'd support frontloading it.

Comment (pichlermarc) [reply content/en/docs/instrumentation/js/manual.md]: We discussed it in the SIG. There were no objections to removing the said methods, so I opened a PR to deprecate them. If it goes through we will remove them in later releases. https://github.com/open-telemetry/opentelemetry-js/pull/3996

The config options of NodeSDK will remain as-is (some logs support will be added in the upcoming release).

Long story short: It's safe to move on with using `@opentelemetry/sdk-node` in the docs :slightly_smiling_face:

Comment (cartermp) [reply content/en/docs/instrumentation/js/manual.md]: That's superb, thanks @pichlermarc!

File: content/en/docs/instrumentation/js/manual.md
Comment (JamieDanielson) [inline content/en/docs/instrumentation/js/manual.md:15]:
```suggestion
On this page you will learn how you can add traces, metrics and logs to your
```

File: content/en/docs/instrumentation/js/manual.md
Comment (JamieDanielson) [inline content/en/docs/instrumentation/js/manual.md:71]:
```suggestion
Create the _library file_ named `dice.ts` (or `dice.js` if you are not using
```

File: content/en/docs/instrumentation/js/manual.md
Comment (JamieDanielson) [inline content/en/docs/instrumentation/js/manual.md:72]:
```suggestion
TypeScript) and add the following code to it:
```

File: content/en/docs/instrumentation/js/manual.md
Comment (JamieDanielson) [inline content/en/docs/instrumentation/js/manual.md:115]:
```suggestion
Create the _app file_ named `app.ts` (or `app.js` if not using TypeScript) and
```

File: content/en/docs/instrumentation/js/manual.md
Comment (JamieDanielson) [inline content/en/docs/instrumentation/js/manual.md:212]:
```suggestion
implementations will be provided to any library that acquires a tracer or meter
```

File: content/en/docs/instrumentation/js/manual.md
Comment (JamieDanielson) [inline content/en/docs/instrumentation/js/manual.md:269]:
```suggestion
For debugging and local development purposes, the following example exports
```

File: content/en/docs/instrumentation/js/manual.md
Comment (JamieDanielson) [inline content/en/docs/instrumentation/js/manual.md:275]:
```suggestion
The example also sets up SDK default attributes, which are 
```

File: content/en/docs/instrumentation/js/manual.md
Comment (JamieDanielson) [inline content/en/docs/instrumentation/js/manual.md:605]:
```suggestion
The API of OpenTelemetry JavaScript exposes two methods that allow you to
```

File: content/en/docs/instrumentation/js/manual.md
Comment (JamieDanielson) [inline content/en/docs/instrumentation/js/manual.md:616]:
```suggestion
care of setting the span and its context active.
```

File: content/en/docs/instrumentation/js/manual.md
Comment (JamieDanielson) [inline content/en/docs/instrumentation/js/manual.md:1264]:
```suggestion
##### Initializing metrics with `sdk-metrics`
```

File: content/en/docs/instrumentation/js/manual.md
Comment (svrnm) [reply content/en/docs/instrumentation/js/manual.md:275]: Since service.name is required and service.version is optional but recommended saying that they are "beneficial" alone is not strong enough.

> The example also sets up SDK default attributes:

maybe we can live without both?

Comment (JamieDanielson) [reply content/en/docs/instrumentation/js/manual.md:275]: Yeah hm... maybe "The example also sets up SDK default attributes, including the required service name" ?

Comment (svrnm) [reply content/en/docs/instrumentation/js/manual.md]: tried a different wording, please take another look and what you think.

**Comment by svrnm** [comment]
@chalin the markdown lint is dropping me some errors I can not manage, as they are related to some hugo-quirks (shortcodes, link fragments, etc.), can you take a look at that?

**Comment by chalin** [comment]
@svrnm - yes, I think I know why. Looking into it now.

**Comment by chalin** [comment]
@svrnm - could you rebase and resolve conflicts first?

**Comment by svrnm** [comment]
> @svrnm - could you rebase and resolve conflicts first?

I don't see a conflict

**Comment by chalin** [comment]
> I don't see a conflict

Try to rebase from `main@HEAD` locally.

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:172]:
```suggestion
<!-- markdownlint-disable -->
<!-- prettier-ignore-start -->
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:186]:
```suggestion
<!-- prettier-ignore-end -->
<!-- markdownlint-restore -->
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:660]:
```suggestion
<!-- prettier-ignore-end -->
<!-- markdownlint-restore -->
```

File: content/en/docs/instrumentation/js/manual.md
Comment (chalin) [inline content/en/docs/instrumentation/js/manual.md:885]:
```suggestion
<!-- prettier-ignore-end -->
<!-- markdownlint-restore -->
```

**Review by chalin — COMMENTED** [review]
@svrnm - you'll need to add a few "ignore" directives for the markdownlint checker. See inline for suggestions.

Regarding the spelling checker reported issues, I'm hoping that #3097 (once merged) will resolve the problems.

**Comment by chalin** [comment]
Rebasing should resolve the check-failure issues.

**Comment by svrnm** [comment]
@JamieDanielson @pichlermarc any more remarks from your site?

**Comment by chalin** [comment]
@svrnm - can you rebase and resolve conflicts?

**Review by JamieDanielson — COMMENTED** [review]
Sorry I just realized I didn't hit submit on this earlier. I added a few more suggestions after another walk-through.

It seems like we are also a bit inconsistent in showing `import opentelemetry from "@opentelemetry/api";` with `opentelemetry.trace.getTracer` and `import { trace } from "@opentelemetry/api"` with `trace.getTracer`. I wasn't able to comment on them all with suggestions because some weren't touched in this PR, so maybe this will be good and we can do another pass-through in a new PR since those changes weren't introduced here?

**Review by JamieDanielson — APPROVED** [review]
This looks great, thanks for all the hard work (and patience 😅 ) on this!

**Pull request merged**
